docs/projects/insertion-sort/_posts/2019-01-22-python.md

```diff
@@ -0,0 +1,218 @@
+---
+title: Insertion Sort in Python
+layout: default
+last-modified: 2019-01-24
+featured-image: insertion-sort-in-python-featured-image.JPEG
+tags: [python, insertion-sort]
+authors:
+  - hmajid2301
+---
+  
+In this article, we'll be tackling Insertion Sort in Python.
+
+## How to Implement the Solution
+
+At this point, let's dig into the code a bit. The following sections break
+down the Insertion Sort in Python functionality.
+
+### Solution
+
+```python
+#!/usr/bin/env python
+import sys
+from itertools import takewhile
+
+
+def insertion_sort(xs):
+    new_xs = []
+    for x in xs:
+        new_xs = insert(x, new_xs)
+    return new_xs
+
+
+def insert(x, xs):
+    left = list(takewhile(lambda i: i < x, xs))
+    right = xs[len(left):] if xs else xs
+    return left + [x] + right
+
+
+def input_list(list_str):
+    return [int(x.strip(" "), 10) for x in list_str.split(',')]
+
+
+def exit_with_error():
+    print('Usage: please provide a list of at least two integers to sort in the format “1, 2, 3, 4, 5”')
+    sys.exit(1)
+
+
+def main(args):
+    try:
+        xs = input_list(args[0])
+        if len(xs) <= 1:
+            exit_with_error()
+        print(insertion_sort(xs))
+    except (IndexError,ValueError):
+        raise
+        exit_with_error()
+
+
+if __name__ == "__main__":
+    main(sys.argv[1:])
+```
+
+### The Main Function
+
+Breaking down this solution bottom up,
+
+```python
+if __name__ == "__main__":
+    main(sys.argv[1:])
+```
+
+This bit of code checks to see if this is the main module run. If it is, it then calls the main
+function and passes user input to it. In this case the user input would be a string of numbers
+like so `"2, 1, 10, 5, 3"` (to sort).
+
+```python
+def main(args):
+    try:
+        xs = input_list(args[0])
+        if len(xs) <= 1:
+            exit_with_error()
+        print(insertion_sort(xs))
+    except (IndexError,ValueError):
+        exit_with_error()
+```
+
+This is the main function of this file. It parses the input, then calls our insertion sort
+function (and prints the results). It also deals with any errors raised.
+
+### Transform Input Parameters
+
+```python
+def input_list(list_str):
+    return [int(x.strip(" "), 10) for x in list_str.split(',')]
+```
+
+This function takes a string like `"2, 1, 10, 5, 3"`, and turns into a list of numbers.
+It does this using a list comprehension. First, we need to convert our string into a
+list `list_str.split(',')` which is a list of strings split by comma (,).
+So our original input string becomes `["2", " 1", " 10", " 5", " 3"]`. Then for each
+element in the list `for x in ...` ,  we do something to it.
+
+In this example we convert it into a decimal integer, `int(x.strip(" "), 10)`. Then `x.strip(" ")`,
```

Comment:
The comma can be shifted up here, so that the code snippet is the subject. 

```suggestion
In this example we convert it into a decimal integer, `int(x.strip(" "), 10)`. Then, `x.strip(" ")`
```